src/models/ClassPotentialR2HDM.cpp

```diff
@@ -743,93 +713,270 @@ std::vector<double> Class_Potential_R2HDM::calc_CT() const
+  if (!CheckRotationMatrix()) // Check whether generically generated rotation
+                              // matrix is proper rotation matrix
+  {
+    throw std::runtime_error("Error in rotation matrix.");
+  }
+
+  MatrixXd HiggsRot(NHiggs, NHiggs);
   for (std::size_t i = 0; i < NHiggs; i++)
   {
     for (std::size_t j = 0; j < NHiggs; j++)
     {
-      for (std::size_t k = 0; k < NHiggs; k++)
+      HiggsRot(i, j) = HiggsRotationMatrix[i][j];
+    }
+  }
+
+  // initialize position indices (new initialization for each point in multiline
+  // files)
+  pos_G1 = -1, pos_G2 = -1, pos_H1 = -1, pos_H2 = -1, pos_G0 = -1, pos_A = -1,
+  pos_H = -1, pos_h = -1;
+
+  // interaction basis
+  // rho1, eta1, rho2, eta2, zeta1, psi1, zeta2, psi2
+  int pos_rho1 = 0, pos_eta1 = 1, pos_rho2 = 2, pos_eta2 = 3, pos_zeta1 = 4,
+      pos_psi1 = 5, pos_zeta2 = 6, pos_psi2 = 7;
+
+  // higgsbasis = {rho1, eta1, rho2, eta2, zeta1, psi1, zeta2, psi2}
+  for (std::size_t i = 0; i < NHiggs;
+       i++) // mass base index i corresponds to mass vector sorted in ascending
+            // mass
+  {
+    // charged submatrices
+    if (std::abs(HiggsRot(i, pos_rho1)) + std::abs(HiggsRot(i, pos_rho2)) >
+        ZeroThreshold) // use that mGpm < mHpm
+    {
+      if (pos_G1 == -1)
       {
-        GaugeBasis[i][j][k] =
-            TripleDeriv.at(i + j * NHiggs + k * NHiggs * NHiggs);
+        pos_G1 = i;
+      }
+      else
+      {
+        pos_H1 = i;
+      }
+    }
+    else if (std::abs(HiggsRot(i, pos_eta1)) + std::abs(HiggsRot(i, pos_eta2)) >
+        ZeroThreshold) // use that mGpm < mHpm
+    {
+      if (pos_G2 == -1)
+      {
+        pos_G2 = i;
+      }
+      else
+      {
+        pos_H2 = i;
+      }
+    }
+    else if (std::abs(HiggsRot(i, pos_zeta1)) + std::abs(HiggsRot(i, pos_zeta2)) >
+        ZeroThreshold) // use that mh < mH
+    {
+      if (pos_h == -1)
+      {
+        pos_h = i;
+      }
+      else
+      {
+        pos_H = i;
+      }
+    }
+    else if (std::abs(HiggsRot(i, pos_psi1)) + std::abs(HiggsRot(i, pos_psi2)) >
+        ZeroThreshold) // use that 0 = mG0 < mA
+    {
+      if (pos_G0 == -1)
+      {
+        pos_G0 = i;
+      }
+      else
+      {
+        pos_A = i;
       }
     }
   }
 
-  MatrixXd HiggsRot(NHiggs, NHiggs);
+  // check if all position indices are set
+  if (pos_G1 == -1 or pos_G2 == -1 or pos_H1 == -1 or pos_H2 == -1 or pos_G0 == -1 or
+      pos_A == -1 or pos_H == -1 or pos_h == -1)
+  {
+    throw std::runtime_error("Error. Not all position indices are set.");
+  }
+
+  // check if all other elements of rotation matrix are zero
+  bool zero_element = false;
   for (std::size_t i = 0; i < NHiggs; i++)
   {
     for (std::size_t j = 0; j < NHiggs; j++)
     {
-      HiggsRot(i, j) = HiggsRotationMatrix[i][j];
+      int ii = int(i);
+      int jj = int(j);
+      if (not((jj == pos_rho1 and (ii == pos_G1 or ii == pos_H1)) or
```

Comment:
This piece of code is a bit confusing, what is it checking for? Can you refactor it a little bit?